src/test/java/org/commcare/formplayer/services/FormDefinitionServiceTest.java

```diff
@@ -162,20 +176,104 @@ public void testGetOrCreateFormDefinitionNewVersionNotEqual() {
         assertNotEquals(formDefinitionV1, formDefinitionV2);
     }
 
-    private Optional<SerializableFormDefinition> getCachedFormDefinition(
+    @Test
+    public void testWriteToLocalStorage() {
+        this.formDefinitionService.writeToLocalStorage(this.formDef);
+        FormDef formDef = getFormDefStorage().getRecordForValue("XMLNS", this.formXmlns);
+        assertThat(formDef).isNotNull();
+    }
+
+    @Test
+    public void testWriteToLocalStorage_FormDefExists() {
+        getFormDefStorage().write(formDef);
+        boolean valueWritten = this.formDefinitionService.writeToLocalStorage(this.formDef);
+        assertThat(valueWritten).isFalse();
+    }
+
+    @Test
+    public void testGetFormDefCaching() {
+        SerializableFormDefinition formDef = this.formDefinitionService.getOrCreateFormDefinition(
+                this.appId, this.formXmlns, this.formVersion, this.formDef
+        );
+        String sessionId = UUID.randomUUID().toString();
+        SerializableFormSession session = new SerializableFormSession(sessionId);
+        session.setFormDefinition(formDef);
+
+        assertThat(getCachedFormDefinition(sessionId)).isEmpty();
+        this.formDefinitionService.getFormDef(session);
+        assertThat(getCachedFormDefinition(sessionId)).isNotEmpty();
+    }
+
+    @Test
+    public void testGetFormDefBrokenSerialization() {
+        SerializableFormDefinition formDef = this.formDefinitionService.getOrCreateFormDefinition(
+                this.appId, this.formXmlns, this.formVersion, this.formDef
+        );
+
+        // put a bad value in `serializedFormDef` to trigger a serialization error
+        ReflectionTestUtils.setField(formDef, "serializedFormDef", "not a form def");
+        SerializableFormSession session = new SerializableFormSession(UUID.randomUUID().toString());
+        session.setFormDefinition(formDef);
+
+        // make sure we're not going to hit the cache
+        assertThat(getCachedFormDefinition(session.getId())).isEmpty();
+
+        // write the formDef to storage so that we can read it back
+        getFormDefStorage().write(this.formDef);
+        assertThat(this.formDef.getID()).isNotNull();
+
+        FormDef reSerializedFormDef = this.formDefinitionService.getFormDef(session);
+        assertThat(reSerializedFormDef.getID()).isEqualTo(this.formDef.getID());
+
+    }
+
+    @Test
+    public void testGetFormDefFromStorageUpdatesSerializableFormDefinition() throws Exception {
```

Comment:
we can probably abstract out the common setup between this and `testGetFormDefBrokenSerialization` into another method.

Reply:
Agreed that there is room for abstraction, but it's one line that is specifically related to setting up a broken serialization. I also added an additional test that relies on this setup. While all three tests could be consolidated into one test with multiple assertions, in HQ we tend to keep assertions in separate tests to make it easy to reason with.

All that to say I don't mind a bit of repetition if it keeps tests relatively simple. I don't see a super obvious abstraction to be made that wouldn't overcomplicate things, so feel free to highlight that or suggest a change if you want. Otherwise I'll likely leave as is.

Reply:
You could extract out the test setup:

```java
public SerializableFormDefinition createBrokenFormDef() {
    SerializableFormDefinition formDef = this.formDefinitionService.getOrCreateFormDefinition(
            this.appId, this.formXmlns, this.formVersion, this.formDef
    );
    ReflectionTestUtils.setField(formDef, "serializedFormDef", "not a form def");

    // this line added based on PR comments below
    formDefinitionRepo.save(formDef);

    SerializableFormSession session = new SerializableFormSession(UUID.randomUUID().toString());
    session.setFormDefinition(formDef);
    getFormDefStorage().write(this.formDef);
    
    return formDef
}

Reply:
Fair point 😅  see 70111ce3e775fe57f4a4bda134179ab20b554d5d